File: solutions/c/high-scores/1/high_scores.c
#include "high_scores.h"

int32_t latest(const int32_t *scores, size_t scores_len){
    return *(scores+scores_len-1);
}

int32_t personal_best(const int32_t *scores, size_t scores_len){
    int32_t best_score=0;
    for (size_t i=0;i<scores_len;i++){
        if (*(scores+i)>best_score) best_score=*(scores+i);
    }
    return best_score;
}

size_t personal_top_three(const int32_t *scores, size_t scores_len,
                          int32_t *output){
    size_t entries=0;
    if (scores_len <3) entries=scores_len;
    else entries=3;
    for (size_t i=0;i<scores_len;i++){
        if (*(scores+i)>=*(output)){
            *(output+2)=*(output+1);
            *(output+1)=*output;
            *output=*(scores+i);
        }
        else if (*(scores+i)>=*(output+1)){
            *(output+2)=*(output+1);
            *(output+1)=*(scores+i);
        }
        else if (*(scores+i)>=*(output+2)){
            *(output+2)=*(scores+i);
        }
    }
    return entries;
}

File: solutions/fortran/high-scores/1/high_scores.f90

module high_scores
  implicit none
contains

function scores(score_list)
  integer, dimension(:) :: score_list
  integer, dimension(size(score_list)) :: scores
  scores=score_list
end function

function latest(score_list)
  integer, dimension(:) :: score_list
  integer :: latest
  latest=score_list(size(score_list))
end function

function personalBest(score_list)
  integer, dimension(:) :: score_list
  integer, dimension(size(score_list)) :: scores
  integer :: personalBest
  scores=sorted(score_list)
  personalBest=scores(1)
end function

function personalTopThree(score_list)
  integer, dimension(:) :: score_list
  integer, dimension(3) :: personalTopThree
  personalTopThree=(/0,0,0/)
  personalTopThree=sorted(score_list)
end function

function sorted(score_list)
  integer, dimension(:) :: score_list
  integer, dimension(size(score_list)) :: sorted
  integer :: i, j, temp

  sorted=score_list
  do i=1, size(sorted)
    do j=i+1, size(sorted)
      if (sorted(i)<sorted(j)) then
        temp=sorted(j)
        sorted(j)=sorted(i)
        sorted(i)=temp
      end if
    end do
  end do

end function

end module
